src/os/linux/mod.rs

```diff
@@ -0,0 +1,165 @@
+mod xdg;
+
+use std::{
+    ffi::{OsStr, OsString},
+    fmt::{self, Display},
+    path::{Path, PathBuf},
+};
+
+#[derive(Debug)]
+pub enum DetectEditorError {
+    NoDefaultEditorSet,
+    FreeDesktopEntryNotFound,
+    FreeDesktopEntryParseError,
+}
+
+impl Display for DetectEditorError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::NoDefaultEditorSet => write!(f, "No default editor is set: xdg-mime queries for \"text/rust\" and \"text/plain\" both failed"),
+            Self::FreeDesktopEntryNotFound => write!(f, "Entry Not Found: xdg-mime returned an entry name that could not be found"),
+            Self::FreeDesktopEntryParseError => write!(f, "Entry Parse Error: xdg-mime returned an entry that could not be parsed"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum OpenFileError {
+    LaunchFailed(bossy::Error),
+}
+
+impl Display for OpenFileError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::LaunchFailed(e) => write!(f, "Launch failed: {}", e),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Application {
+    exec_command: OsString,
+    icon: Option<OsString>,
+    xdg_entry_path: PathBuf,
+}
+
+impl Application {
+    pub fn detect_editor() -> Result<Self, DetectEditorError> {
+        let entry = {
+            // Try a rust code editor, then a plain text editor, if neither are available,
+            // then return an error.
+            if let Some(rust_entry) = xdg::query_mime_entry("text/rust") {
+                rust_entry
+            } else if let Some(plain_entry) = xdg::query_mime_entry("text/plain") {
+                plain_entry
+            } else {
+                return Err(DetectEditorError::NoDefaultEditorSet);
+            }
+        };
+
+        for dir in xdg::get_xdg_data_dirs() {
```

Comment:
You can use [`find_map`](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.find_map) to simplify the logic here:
```rust
let path = xdg::get_xdg_data_dirs()
    .into_iter()
    .find_map(|dir| {
        xdg::find_entry_in_dir(&dir.join("applications"), &entry)
    })
    .ok_or(DetectEditorError::FreeDesktopEntryNotFound)?;
let parsed = xdg::parse(&path)
    .ok_or(DetectEditorError::FreeDesktopEntryParseError)?;
Ok(Self {
    exec_command: parsed
        .section("Desktop Entry")
        .attr("Exec")
        .ok_or(DetectEditorError::FreeDesktopEntryParseError)?
        .into(),
    icon: parsed
        .section("Desktop Entry")
        .attr("Icon")
        .map(Into::into),
    xdg_entry_path: path,
})
```